explore/explore_internal.h

```diff
@@ -294,56 +298,31 @@ namespace exploration
     return sample_after_normalizing(seed, pdf_first, pdf_last, chosen_index, pdf_category());
   }
 
-  template<typename PdfIt, typename InputScoreIt, typename OutputIt>
-  int sample_after_normalizing(uint64_t seed,
-      PdfIt pdf_first, PdfIt pdf_last, std::random_access_iterator_tag pdf_category,
-      InputScoreIt scores_first, InputScoreIt scores_last, std::random_access_iterator_tag scores_category,
-      OutputIt ranking_first, OutputIt ranking_last, std::random_access_iterator_tag ranking_category)
+  template<typename ActionIt>
+  int swap_chosen(ActionIt action_first, ActionIt action_last, std::forward_iterator_tag action_category, uint32_t chosen_index)
   {
-    if (pdf_last < pdf_first || ranking_last < ranking_first)
+    if ( action_last < action_first )
```

Comment:
[nit] Looks like all E_EXPLORATION_BAD_RANGE checks (304-312) can be collapsed into: if (action_first + chosen_index >= action_last)

Reply:
Had to revert this change since not all iterators have all comparison operators.  Moved back to conservative implementation.